File: Tests/iOS Tests/PNPublishTests.m
//
//  PNPublishTests.m
//  PubNub Tests
//
//  Created by Jordan Zucker on 6/15/15.
//
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import <PubNub/PubNub.h>

#import "PNBasicClientTestCase.h"

@interface PNPublishTests : PNBasicClientTestCase
@end

@implementation PNPublishTests

- (BOOL)isRecording {
    return NO;
}

- (void)testSimplePublish {
    NSString *uniqueChannel = @"2EC925F0-B996-47A4-AF54-A605E1A9AEBA";
    [self performVerifiedPublish:@"test" onChannel:uniqueChannel
                  withAssertions:^(PNPublishStatus *status) {
        XCTAssertNotNil(status);
        XCTAssertEqual(status.category, PNAcknowledgmentCategory);
        XCTAssertEqual(status.operation, PNPublishOperation);
        XCTAssertEqual(status.statusCode, 200);
        XCTAssertFalse(status.isError);
        NSLog(@"status.data.information: %@", status.data.information);
        NSLog(@"status.data.timeToken: %@", status.data.timetoken);
        XCTAssertEqualObjects(status.data.information, @"Sent");
        XCTAssertEqualObjects(status.data.timetoken, @"14345797352002658");
    }];
}

- (void)testPublishNilMessage {
    NSString *uniqueChannel = @"2EC925F0-B996-47A4-AF54-A605E1A9AEBA";
    [self performVerifiedPublish:nil onChannel:uniqueChannel
                  withAssertions:^(PNPublishStatus *status) {
        XCTAssertNotNil(status);
        XCTAssertEqual(status.category, PNBadRequestCategory);
        XCTAssertEqual(status.operation, PNPublishOperation);
        XCTAssertEqual(status.statusCode, 400);
        XCTAssertTrue(status.isError);
        NSLog(@"status.data.information: %@", status.data.information);
        NSLog(@"status.data.timeToken: %@", status.data.timetoken);
        XCTAssertNil(status.data.information);
        XCTAssertNil(status.data.timetoken);
    }];
}

- (void)testPublishDictionary {
    NSString *uniqueChannel = @"2EC925F0-B996-47A4-AF54-A605E1A9AEBA";
    [self performVerifiedPublish:@{@"test" : @"test"} onChannel:uniqueChannel
                  withAssertions:^(PNPublishStatus *status) {
        XCTAssertNotNil(status);
        XCTAssertEqual(status.category, PNAcknowledgmentCategory);
        XCTAssertEqual(status.operation, PNPublishOperation);
        XCTAssertEqual(status.statusCode, 200);
        XCTAssertFalse(status.isError);
        NSLog(@"status.data.information: %@", status.data.information);
        NSLog(@"status.data.timeToken: %@", status.data.timetoken);
        XCTAssertEqualObjects(status.data.information, @"Sent");
        XCTAssertEqualObjects(status.data.timetoken, @"14345797350907256");
    }];
}

- (void)performVerifiedPublish:(id)message onChannel:(NSString *)channel withAssertions:(PNPublishCompletionBlock)verificationBlock {
    XCTestExpectation *networkExpectation = [self expectationWithDescription:@"network"];
    [self.client publish:message toChannel:channel
          withCompletion:^(PNPublishStatus *status) {
        verificationBlock(status);
        [networkExpectation fulfill];
    }];
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

@end

File: Tests/iOS Tests-Swift/PNPublishTests.swift
//
//  PNPublishTests.swift
//  PubNub Tests
//
//  Created by Jordan Zucker on 7/28/15.
//
//

import UIKit
import XCTest

class PNPublishTests: PNBasicClientTestCase {

    override func isRecording() -> Bool {
        return false
    }
    
    let publishTestsChannelName = "2EC925F0-B996-47A4-AF54-A605E1A9AEBA"
    
    func testSimplePublish() {
        self.performVerifiedPublish("test", onChannel: publishTestsChannelName) { (status) -> Void in
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information, "Sent")
            XCTAssertEqual(status.data.timetoken, NSNumber(longLong: 14355311066264140))
        }
    }
    
    func testPublishDictionary() {
        self.performVerifiedPublish(["test" : "test"], onChannel: publishTestsChannelName) { (status) -> Void in
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information, "Sent")
            XCTAssertEqual(status.data.timetoken, NSNumber(longLong: 14355311062532489))
        }
    }

}

extension PNPublishTests {
    func performVerifiedPublish(message: AnyObject, onChannel:String, withAssertions: PNPublishCompletionBlock) {
        let networkExpectation = self.expectationWithDescription("network")
        client.publish(message, toChannel: onChannel) { (status) -> Void in
            withAssertions(status)
            networkExpectation.fulfill()
        }
        waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
            XCTAssertNil(error, "Encountered error with publish call")
        })
    }
}
